routing/grandcentral/server.go

```diff
@@ -42,91 +42,53 @@ func (s *Server) HandleRequest(ctx context.Context, p peer.ID, req *dhtpb.Messag
 func (s *Server) handleMessage(
 	ctx context.Context, p peer.ID, req *dhtpb.Message) (peer.ID, *dhtpb.Message) {
 
-	//  FIXME threw everything into this switch statement to get things going.
-	//  Once each operation is well-defined, extract pluggable backend so any
-	//  database may be used.
+	log.EventBegin(ctx, "routingMessageReceived", req, p, s.local).Done() // TODO may need to differentiate between local and remote
 
 	var response = dhtpb.NewMessage(req.GetType(), req.GetKey(), req.GetClusterLevel())
 	switch req.GetType() {
 
 	case dhtpb.Message_GET_VALUE:
-		dskey := util.Key(req.GetKey()).DsKey()
-		val, err := s.datastore.Get(dskey)
+		rawRecord, err := getRoutingRecord(s.routingBackend, util.Key(req.GetKey()))
 		if err != nil {
-			log.Error(errors.Wrap(err))
-			return "", nil
-		}
-		rawRecord, ok := val.([]byte)
-		if !ok {
-			log.Errorf("datastore had non byte-slice value for %v", dskey)
-			return "", nil
-		}
-		if err := proto.Unmarshal(rawRecord, response.Record); err != nil {
-			log.Error("failed to unmarshal dht record from datastore")
 			return "", nil
 		}
+		response.Record = rawRecord
 		// TODO before merging: if we know any providers for the requested value, return those.
 		return p, response
 
 	case dhtpb.Message_PUT_VALUE:
 		// TODO before merging: verifyRecord(req.GetRecord())
-		data, err := proto.Marshal(req.GetRecord())
-		if err != nil {
-			log.Error(err)
-			return "", nil
-		}
-		dskey := util.Key(req.GetKey()).DsKey()
-		if err := s.datastore.Put(dskey, data); err != nil {
-			log.Error(err)
-			return "", nil
-		}
+		putRoutingRecord(s.routingBackend, util.Key(req.GetKey()), req.GetRecord())
 		return p, req // TODO before merging: verify that we should return record
 
 	case dhtpb.Message_FIND_NODE:
 		p := s.peerstore.PeerInfo(peer.ID(req.GetKey()))
-		response.CloserPeers = dhtpb.PeerInfosToPBPeers(s.dialer, []peer.PeerInfo{p})
+		pri := []dhtpb.PeerRoutingInfo{
+			dhtpb.PeerRoutingInfo{
+				PeerInfo: p,
+				// Connectedness: TODO
+			},
+		}
+		response.CloserPeers = dhtpb.PeerRoutingInfosToPBPeers(pri)
```

Comment:
This means that the router's internal cache is the only way to get peer addresses. This may be fine, given that getting addresses isnt a problem atm, but it certainly should be described in some high level description of how GCR works. I would've expected addresses to also be stored in the datastore.